src/Autofac.Extensions.DependencyInjection/AutofacServiceProvider.cs

```diff
@@ -119,5 +121,22 @@ public void Dispose()
             this.Dispose(true);
             GC.SuppressFinalize(this);
         }
+
+        /// <summary>
+        /// Performs a dispose operation asynchronously.
+        /// </summary>
+        [SuppressMessage(
+            "Usage",
+            "CA1816:Dispose methods should call SuppressFinalize",
+            Justification = "DisposeAsync should also call SuppressFinalize (see various .NET internal implementations).")]
+        public async ValueTask DisposeAsync()
+        {
+            if (!this._disposed)
+            {
+                await this._lifetimeScope.DisposeAsync();
+                GC.SuppressFinalize(this);
+                this._disposed = true;
```

Comment:
Doesn't it make sense to immediately call `this._disposed = true`?

```csharp
this._disposed = true;
await this._lifetimeScope.DisposeAsync();
GC.SuppressFinalize(this);
```

Reply:
Possibly; I just followed the pattern of the normal Dispose call that already exists, which sets ``__diposed`` afterwards.  If we change one we should probably change the other.

Reply:
Not sure which is the better solution I am finding different in several Microsoft repositories.

For instance the (MemoryCache)[https://github.com/aspnet/Extensions/blob/master/src/Caching/Memory/src/MemoryCache.cs#L481] does it like we are currently doing it (kind of). 

The `ServiceProviderEngine` does it (differently)[https://github.com/aspnet/Extensions/blob/master/src/DependencyInjection/DI/src/ServiceLookup/ServiceProviderEngine.cs#L68] though.

Reply:
An effect of changing it to before (that I can think of) would be that if an exception threw from the nested DisposeAsync, calling DisposeAsync again would not attempt a nested DisposeAsync (which might be correct behaviour).

Thinking about it, it would also reduce the risk of someone calling AsyncDispose, not awaiting it, and then calling Dispose manually, so I think I might move it (and I'll update the existing Dispose method to match).

Thanks for opening the discussion!

Reply:
> Thinking about it, it would also reduce the risk of someone calling AsyncDispose, not awaiting it, and then calling Dispose manually, so I think I might move it (and I'll update the existing Dispose method to match).

That's exactly a use-case I've thought about here. Also someone could call regular `Dispose` in  an asnyc context without waiting for it to completed where we would end up with the same problem.

Reply:
I've reordered dispose for the service scope and service provider.